File: character-sheet-companion-server/src/listeners/rolls/rollHitDieListener.ts
import { Server, Socket } from "socket.io";
import {
  getFoundrySocketFromRoom,
  getIosSocketFromRoom,
} from "../../utilities/SocketUtilities";
import { SOCKET_EVENTS } from "../../constants/events";
import { RollHitDie } from "../../interfaces/rolls/RollHitDie";

export const rollHitDieListener = (io: Server, socket: Socket): void => {
  const requestFoundryHitDieRoll = (rollHitDie: RollHitDie) => {
    const foundrySocket = getFoundrySocketFromRoom(io, socket);
    if (foundrySocket) {
      foundrySocket.emit(
        SOCKET_EVENTS.SERVER.ROLL.REQUEST.REQUEST_FOUNDRY_HIT_DIE_ROLL,
        rollHitDie,
        socket.id
      );
    }
  };

  const receiveFoundryHitDieRoll = (
    rollHitDie: RollHitDie,
    iosSocketId: string
  ) => {
    const iosSocket = getIosSocketFromRoom(io, socket, iosSocketId);
    if (iosSocket) {
      iosSocket.emit(
        SOCKET_EVENTS.SERVER.ROLL.SEND.SEND_FOUNDRY_ABILITY_ROLL,
        JSON.stringify(rollHitDie)
      );
    }
  };

  socket.on(
    SOCKET_EVENTS.IOS.ROLL.REQUEST_FOUNDRY_HIT_DIE_ROLL,
    requestFoundryHitDieRoll
  );
  socket.on(
    SOCKET_EVENTS.FOUNDRY.ROLL.SEND_HIT_DIE_ROLL,
    receiveFoundryHitDieRoll
  );
};

File: CharacterSheetCompanion/CharacterSheetCompanion/Services/SocketListeners/RollListeners/RollHitDieListener.swift
//
//  InitiativeListener.swift
//  CharacterSheetCompanion
//
//  Created by Zachary Sanders on 8/2/21.
//

import Foundation
import SocketIO

class RollHitDieListener: SocketListener {
    let socket: SocketIOClient

    private let jsonEncoder = JSONEncoder()
    private var hitDieCallback: ((HitDieRollModel) -> Void)?

    init(socket: SocketIOClient) {
        self.socket = socket
    }

    func addSocketHandlers() {
        socket.on(SocketEvents.SERVER.ROLL.SEND.SEND_HIT_DIE_ROLL) { data, _ in
            do {
                try self.hitDieCallback?(SocketListenerUtility.parseSocketEventData(data))
            } catch let FoundryJSONError.errorMessage(errorMessage) {
                print(errorMessage)
            } catch {
                print(error)
            }
        }
    }

    func rollHitDie(hitDieRoll: HitDieRollModel, completionHandler: @escaping (HitDieRollModel) -> Void) {
        do {
            let jsonData = try jsonEncoder.encode(hitDieRoll)
            if let json = String(data: jsonData, encoding: .utf8) {
                socket.emit(SocketEvents.IOS.ROLL.REQUEST_FOUNDRY_HIT_DIE_ROLL, json)
                hitDieCallback = completionHandler
            }
        } catch {}
    }
}

extension SocketEvents.IOS.ROLL {
    static let REQUEST_FOUNDRY_HIT_DIE_ROLL = "ios:requestFoundryHitDieRoll"
}

extension SocketEvents.SERVER.ROLL.SEND {
    static let SEND_HIT_DIE_ROLL = "foundry:sendHitDieRoll"
}
